lib/spack/spack/cmd/module.py

```diff
@@ -23,135 +23,233 @@
+# - mtype : the type of the module file
+# - specs : the list of specs to be processed
+# - args : namespace containing the parsed command line arguments
+callbacks = {}
 
-    find_parser.add_argument(
-        '-r', '--dependencies', action='store_true',
-        dest='recurse_dependencies',
-        help='Recursively traverse dependencies for modules to load.')
 
-    find_parser.add_argument(
-        '-s', '--shell', action='store_true', dest='shell',
-        help='Generate shell script (instead of input for module command)')
+def subcommand(subparser_name):
+    """Registers a function in the callbacks dictionary"""
+    def decorator(callback):
+        callbacks[subparser_name] = callback
+        return callback
+    return decorator
 
-    find_parser.add_argument(
-        '-p', '--prefix', dest='prefix',
-        help='Prepend to module names when issuing module load commands')
 
-    find_parser.add_argument(
-        'spec', nargs='+',
-        help='spec to find a module file for.')
-
-
-def module_find(mtype, flags, spec_array):
-    """Look at all installed packages and see if the spec provided
-       matches any.  If it does, check whether there is a module file
-       of type <mtype> there, and print out the name that the user
-       should type to use that package's module.
-    prefix:
-        Prepend this to module names when issuing "module load" commands.
-        Some systems seem to need it.
+def setup_parser(subparser):
+    sp = subparser.add_subparsers(metavar='SUBCOMMAND', dest='subparser_name')
+
+    # spack module refresh
+    refresh_parser = sp.add_parser('refresh', help='Regenerate module files')
+    refresh_parser.add_argument(
+        '--delete-tree',
+        help='Delete the module file tree before refresh',
+        action='store_true'
+    )
+    arguments.add_common_arguments(
+        refresh_parser, ['constraint', 'module_type', 'yes_to_all']
+    )
+
+    # spack module find
+    find_parser = sp.add_parser('find', help='Find module files for packages')
+    arguments.add_common_arguments(find_parser, ['constraint', 'module_type'])
+
+    # spack module rm
+    rm_parser = sp.add_parser('rm', help='Remove module files')
+    arguments.add_common_arguments(
+        rm_parser, ['constraint', 'module_type', 'yes_to_all']
+    )
+
+    # spack module loads
+    loads_parser = sp.add_parser(
+        'loads',
+        help='Prompt the list of modules associated with a constraint'
+    )
+    loads_parser.add_argument(
+        '--input-only', action='store_false', dest='shell',
+        help='Generate input for module command (instead of a shell script)'
+    )
+    loads_parser.add_argument(
+        '-p', '--prefix', dest='prefix', default='',
+        help='Prepend to module names when issuing module load commands'
+    )
+    arguments.add_common_arguments(
+        loads_parser, ['constraint', 'module_type', 'recurse_dependencies']
+    )
+
+
+class MultipleMatches(Exception):
```

Comment:
These always get caught in this so it's probably ok, but why not make them extend `SpackException` so they can be handled gracefully if they ever get out?